File: backend/scanner/services/supabase_client.py
"""
Supabase client and utilities for backend integration.
Handles authentication, database operations, and search history.
"""
import os
import json
from supabase import create_client, Client
import logging

logger = logging.getLogger(__name__)


class SupabaseClient:
    """Singleton Supabase client for the application."""
    
    _instance = None
    _client: Client = None
    _db_client: Client = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(SupabaseClient, cls).__new__(cls)
            cls._instance._initialize()
        return cls._instance
    
    def _initialize(self):
        """Initialize Supabase client with credentials from environment."""
        try:
            supabase_url = os.getenv('SUPABASE_URL')
            supabase_anon_key = os.getenv('SUPABASE_ANON_KEY')
            supabase_service_role_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
            
            if not supabase_url or not supabase_anon_key:
                raise ValueError(
                    "Missing SUPABASE_URL or SUPABASE_ANON_KEY in environment"
                )
            
            # Client for auth operations (signup/login/token verification)
            self._client = create_client(supabase_url, supabase_anon_key)

            # Client for database operations from backend.
            # Service role bypasses RLS and we enforce user scoping in queries.
            db_key = supabase_service_role_key or supabase_anon_key
            self._db_client = create_client(supabase_url, db_key)

            logger.info("Supabase clients initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize Supabase client: {e}")
            raise
    
    @property
    def client(self) -> Client:
        """Get Supabase auth client instance."""
        if self._client is None:
            self._initialize()
        return self._client

    @property
    def db_client(self) -> Client:
        """Get Supabase database client instance."""
        if self._db_client is None:
            self._initialize()
        return self._db_client
    
    def verify_token(self, token: str) -> dict:
        """
        Verify JWT token and get user info.
        
        Args:
            token: JWT token from request header
            
        Returns:
            User data dict with user_id, email, etc.
        """
        try:
            user = self.client.auth.get_user(token)
            return {
                'user_id': user.user.id,
                'email': user.user.email,
                'user_metadata': user.user.user_metadata or {}
            }
        except Exception as e:
            logger.error(f"Token verification failed: {e}")
            raise ValueError(f"Invalid token: {str(e)}")
    
    def register_user(self, email: str, password: str) -> dict:
        """
        Register a new user with email and password.
        
        Args:
            email: User email
            password: User password
            
        Returns:
            User data and session
        """
        try:
            response = self.client.auth.sign_up({
                "email": email,
                "password": password
            })
            return {
                'user': {
                    'user_id': response.user.id,
                    'email': response.user.email
                },
                'session': {
                    'access_token': response.session.access_token if response.session else None,
                    'refresh_token': response.session.refresh_token if response.session else None
                }
            }
        except Exception as e:
            logger.error(f"Registration failed: {e}")
            raise
    
    def login_user(self, email: str, password: str) -> dict:
        """
        Login user with email and password.
        
        Args:
            email: User email
            password: User password
            
        Returns:
            User data and session tokens
        """
        try:
            response = self.client.auth.sign_in_with_password({
                "email": email,
                "password": password
            })
            return {
                'user': {
                    'user_id': response.user.id,
                    'email': response.user.email
                },
                'session': {
                    'access_token': response.session.access_token,
                    'refresh_token': response.session.refresh_token
                }
            }
        except Exception as e:
            logger.error(f"Login failed: {e}")
            raise
    
    def add_scan_report(self, user_id: str, repo_url: str, repo_name: str) -> dict:
        """
        Create a new scan report for a user.
        
        Args:
            user_id: Supabase user ID
            repo_url: GitHub repository URL
            repo_name: Repository name
            
        Returns:
            Created scan report with ID
        """
        try:
            data = {
                'user_id': user_id,
                'repo_url': repo_url,
                'repo_name': repo_name,
                'scan_status': 'pending'
            }
            
            response = self.db_client.table('scan_reports').insert(data).execute()
            return response.data[0] if response.data else None
        except Exception as e:
            logger.error(f"Failed to add scan report: {e}")
            raise
    
    def update_scan_report(self, scan_id: str, user_id: str, **updates) -> dict:
        """
        Update an existing scan report.
        
        Args:
            scan_id: Scan report ID
            user_id: User ID (for verification)
            **updates: Fields to update
            
        Returns:
            Updated scan report
        """
        try:
            response = (
                self.db_client.table('scan_reports')
                .update(updates)
                .eq('id', scan_id)
                .eq('user_id', user_id)
                .execute()
            )
            return response.data[0] if response.data else None
        except Exception as e:
            logger.error(f"Failed to update scan report: {e}")
            raise
    
    def get_user_scan_reports(self, user_id: str, limit: int = 50) -> list:
        """
        Get all scan reports for a user.
        
        Args:
            user_id: Supabase user ID
            limit: Max number of reports to fetch
            
        Returns:
            List of scan reports
        """
        try:
            response = (
                self.db_client.table('scan_reports')
                .select('*')
                .eq('user_id', user_id)
                .order('created_at', desc=True)
                .limit(limit)
                .execute()
            )
            return response.data if response.data else []
        except Exception as e:
            logger.error(f"Failed to get scan reports: {e}")
            raise
    
    def get_scan_report(self, scan_id: str, user_id: str) -> dict:
        """
        Get a specific scan report for a user.
        
        Args:
            scan_id: Scan report ID
            user_id: User ID (for verification)
            
        Returns:
            Scan report data
        """
        try:
            response = (
                self.db_client.table('scan_reports')
                .select('*')
                .eq('id', scan_id)
                .eq('user_id', user_id)
                .single()
                .execute()
            )
            return response.data if response.data else None
        except Exception as e:
            logger.error(f"Failed to get scan report: {e}")
            raise
    
    def add_search_history(self, user_id: str, search_query: str, search_result: dict = None) -> dict:
        """
        Add entry to user's search history. Automatically maintains last 10 entries via trigger.
        
        Args:
            user_id: Supabase user ID
            search_query: Search query/GitHub URL
            search_result: Search result JSON
            
        Returns:
            Created history entry
        """
        try:
            data = {
                'user_id': user_id,
                'search_query': search_query,
                'search_result': search_result or {}
            }
            
            response = self.db_client.table('search_history').insert(data).execute()
            return response.data[0] if response.data else None
        except Exception as e:
            logger.error(f"Failed to add search history: {e}")
            raise
    
    def get_search_history(self, user_id: str, limit: int = 10) -> list:
        """
        Get user's search history (last N entries).
        
        Args:
            user_id: Supabase user ID
            limit: Max number of entries (default 10)
            
        Returns:
            List of search history entries
        """
        try:
            response = (
                self.db_client.table('search_history')
                .select('*')
                .eq('user_id', user_id)
                .order('created_at', desc=True)
                .limit(limit)
                .execute()
            )
            return response.data if response.data else []
        except Exception as e:
            logger.error(f"Failed to get search history: {e}")
            raise


# Global Supabase client instance
supabase = SupabaseClient()

File: frontend/src/services/supabaseClient.js
/**
 * Supabase client initialization for frontend.
 * Handles all authentication and database operations.
 */
import { createClient } from '@supabase/supabase-js';

// Get credentials from environment variables
const SUPABASE_URL = import.meta.env.VITE_SUPABASE_URL;
const SUPABASE_ANON_KEY = import.meta.env.VITE_SUPABASE_ANON_KEY;

if (!SUPABASE_URL || !SUPABASE_ANON_KEY) {
  console.error(
    'Missing VITE_SUPABASE_URL or VITE_SUPABASE_ANON_KEY in environment variables'
  );
}

// Create Supabase client
export const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

function notifyAuthChanged() {
  window.dispatchEvent(new Event('auth-changed'));
}

async function parseErrorResponse(response, fallbackMessage) {
  const contentType = response.headers.get('content-type') || '';

  if (contentType.includes('application/json')) {
    const error = await response.json();
    return error.error || error.message || fallbackMessage;
  }

  const text = await response.text();
  if (text && text.toLowerCase().includes('<!doctype')) {
    return `${fallbackMessage} (server returned HTML error page)`;
  }

  return fallbackMessage;
}

/**
 * Authentication service class
 */
export class AuthService {
  /**
   * Register a new user
   * @param {string} email - User email
   * @param {string} password - User password
   * @returns {Promise<{user, session}>}
   */
  static async register(email, password) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/auth/register/`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ email, password }),
        }
      );

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.error || 'Registration failed');
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }

  /**
   * Login user
   * @param {string} email - User email
   * @param {string} password - User password
   * @returns {Promise<{user, session}>}
   */
  static async login(email, password) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/auth/login/`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ email, password }),
        }
      );

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.error || 'Login failed');
      }

      const data = await response.json();
      
      // Store tokens in localStorage
      if (data.session?.access_token) {
        localStorage.setItem('access_token', data.session.access_token);
        localStorage.setItem('refresh_token', data.session.refresh_token);
        localStorage.setItem('user', JSON.stringify(data.user));
        notifyAuthChanged();
      }

      return data;
    } catch (error) {
      throw error;
    }
  }

  /**
   * Logout user
   */
  static async logout() {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/auth/logout/`,
        {
          method: 'POST',
          headers: this.getAuthHeaders(),
        }
      );

      // Clear local storage
      localStorage.removeItem('access_token');
      localStorage.removeItem('refresh_token');
      localStorage.removeItem('user');
      notifyAuthChanged();

      return await response.json();
    } catch (error) {
      console.error('Logout error:', error);
      // Force clear even if request fails
      localStorage.removeItem('access_token');
      localStorage.removeItem('refresh_token');
      localStorage.removeItem('user');
      notifyAuthChanged();
    }
  }

  /**
   * Get current user from local storage
   * @returns {object|null}
   */
  static getCurrentUser() {
    const user = localStorage.getItem('user');
    return user ? JSON.parse(user) : null;
  }

  /**
   * Get access token
   * @returns {string|null}
   */
  static getAccessToken() {
    return localStorage.getItem('access_token');
  }

  /**
   * Refresh access token
   * @param {string} refreshToken - Refresh token
   * @returns {Promise<{access_token, refresh_token}>}
   */
  static async refreshToken(refreshToken) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/auth/refresh/`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ refresh_token: refreshToken }),
        }
      );

      if (!response.ok) {
        throw new Error('Token refresh failed');
      }

      const data = await response.json();
      
      if (data.session?.access_token) {
        localStorage.setItem('access_token', data.session.access_token);
        localStorage.setItem('refresh_token', data.session.refresh_token);
        notifyAuthChanged();
      }

      return data.session;
    } catch (error) {
      // Clear tokens if refresh fails
      localStorage.removeItem('access_token');
      localStorage.removeItem('refresh_token');
      throw error;
    }
  }

  /**
   * Get authorization headers with Bearer token
   * @returns {object}
   */
  static getAuthHeaders() {
    const token = this.getAccessToken();
    return {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    };
  }
}

/**
 * Scan service - handles scan operations
 */
export class ScanService {
  /**
   * Start a new scan
   * @param {string} repoUrl - GitHub repository URL
   * @returns {Promise<{scan_id, status}>}
   */
  static async startScan(repoUrl) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/scan/`,
        {
          method: 'POST',
          headers: AuthService.getAuthHeaders(),
          body: JSON.stringify({ repo_url: repoUrl }),
        }
      );

      if (!response.ok) {
        const message = await parseErrorResponse(response, 'Failed to start scan');
        throw new Error(message);
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }

  /**
   * Get scan status
   * @param {string} scanId - Scan ID
   * @returns {Promise<object>}
   */
  static async getScanStatus(scanId) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/scan-status/${scanId}/`,
        {
          headers: AuthService.getAuthHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error('Failed to get scan status');
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }

  /**
   * Get all scan reports for user
   * @param {number} limit - Max number of reports
   * @returns {Promise<array>}
   */
  static async getReports(limit = 50) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/reports/?limit=${limit}`,
        {
          headers: AuthService.getAuthHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error('Failed to get reports');
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }

  /**
   * Get specific report
   * @param {string} reportId - Report ID
   * @returns {Promise<object>}
   */
  static async getReport(reportId) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/reports/${reportId}/`,
        {
          headers: AuthService.getAuthHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error('Failed to get report');
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }
}

/**
 * Search history service
 */
export class HistoryService {
  /**
   * Get search history
   * @param {number} limit - Max number of entries (default 10)
   * @returns {Promise<array>}
   */
  static async getHistory(limit = 10) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/history/?limit=${limit}`,
        {
          headers: AuthService.getAuthHeaders(),
        }
      );

      if (!response.ok) {
        const message = await parseErrorResponse(response, 'Failed to get search history');
        throw new Error(message);
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }

  /**
   * Delete search history entry
   * @param {string} entryId - History entry ID
   * @returns {Promise<object>}
   */
  static async deleteEntry(entryId) {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/history/${entryId}/`,
        {
          method: 'DELETE',
          headers: AuthService.getAuthHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error('Failed to delete history entry');
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }

  /**
   * Clear all search history
   * @returns {Promise<object>}
   */
  static async clearHistory() {
    try {
      const response = await fetch(
        `${import.meta.env.VITE_API_URL}/api/history/`,
        {
          method: 'DELETE',
          headers: AuthService.getAuthHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error('Failed to clear search history');
      }

      return await response.json();
    } catch (error) {
      throw error;
    }
  }
}

export default supabase;
